runtime/gc_vlhgc/CopyForwardScheme.cpp

```diff
@@ -1132,120 +1133,133 @@ MM_CopyForwardScheme::reserveMemoryForCopy(MM_EnvironmentVLHGC *env, J9Object *o
+			j9tty_printf(PORTLIB, "!", addrBase);
 		}
-	}
-
-	/* if objectReserveSizeInBytes is larger then the size we failed to allocate in past during this (aborted) collect, do not bother attempting again */
-	if (NULL == copyCache && (objectReserveSizeInBytes < copyForwardCompactGroup->_failedAllocateSize)) {
-		Assert_MM_true(NULL == copyForwardCompactGroup->_copyCache);
-	
-		/* The copy cache was null or did not have enough room */
+	} else {
 		bool allocateResult = false;
 		MM_LightweightNonReentrantLock *listLock = NULL;
-	
-		if(_minCacheSize < minimumRequiredCacheSize) {
-			addrBase = reserveMemoryForObject(env, compactGroup, minimumRequiredCacheSize, &listLock);
-	
-			if(NULL != addrBase) {
-				addrTop = (void *)(((U_8 *)addrBase) + minimumRequiredCacheSize);
-				if(_cacheTracingEnabled) {
-					PORT_ACCESS_FROM_ENVIRONMENT(env);
-					j9tty_printf(PORTLIB, "Cache specific size %zu allocated: %p %p\n", minimumRequiredCacheSize, addrBase, addrTop);
-				}
+		if (objectReserveSizeInBytes < copyForwardCompactGroup->_failedAllocateSize) {
+			/* try to use TLH remainder from previous discard */
+			if (((uintptr_t)copyForwardCompactGroup->_TLHRemainderTop - (uintptr_t)copyForwardCompactGroup->_TLHRemainderBase) >= minimumRequiredCacheSize) {
 				allocateResult = true;
-			} else {
-				/* failed to allocate - set the threshold to short-circuit future alloc attempts */
-				copyForwardCompactGroup->_failedAllocateSize = objectReserveSizeInBytes;
-			}
-		} else {
-			UDATA desiredCacheSize = getDesiredCopyCacheSize(env, compactGroup);
-			if(reserveMemoryForCache(env, compactGroup, desiredCacheSize, &addrBase, &addrTop, &listLock)) {
-				if(_cacheTracingEnabled) {
+				addrBase = copyForwardCompactGroup->_TLHRemainderBase;
+				addrTop = copyForwardCompactGroup->_TLHRemainderTop;
+				Assert_MM_true(NULL != copyForwardCompactGroup->_TLHRemainderBase);
+				Assert_MM_true(NULL != copyForwardCompactGroup->_TLHRemainderTop);
+				copyForwardCompactGroup->resetTLHRemainder();
+
+				uintptr_t sublistCount = _reservedRegionList[compactGroup]._sublistCount;
+				uintptr_t sublistIndex = env->getWorkerID() % sublistCount;
+				MM_ReservedRegionListHeader::Sublist *regionList = &_reservedRegionList[compactGroup]._sublists[sublistIndex];
+				listLock = &regionList->_lock;
+
+				if (_cacheTracingEnabled) {
 					PORT_ACCESS_FROM_ENVIRONMENT(env);
-					j9tty_printf(PORTLIB, "Cache TLH %zu allocated: %p %p\n", ((UDATA)addrTop) - ((UDATA)addrBase), addrBase, addrTop);
+					j9tty_printf(PORTLIB, "Cache use TLH remainder env=%p, minimumRequiredCacheSize=%zu remainder: %p %p\n",
+									env, minimumRequiredCacheSize, addrBase, addrTop);
+				}
+			} else if (_extensions->tlhSurvivorDiscardThreshold < minimumRequiredCacheSize) {
+				addrBase = reserveMemoryForObject(env, compactGroup, minimumRequiredCacheSize, &listLock);
+
+				if (NULL != addrBase) {
+					addrTop = (void *)(((U_8 *)addrBase) + minimumRequiredCacheSize);
```

Comment:
use uintptr_t like elsewhere

Reply:
seems like this is old line of code, but just got indented
in fact, this file still uses U_8 and UDATA for the most part, so you don't really change it, if you did not touch it, after all